File: textbook/python/06-pact/04_envelopes.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — PACT / Operating Envelopes
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Create and intersect operating envelopes with monotonic tightening
# LEVEL: Intermediate
# PARITY: Full — Rust has equivalent envelope intersection semantics
# VALIDATES: RoleEnvelope, TaskEnvelope, intersect_envelopes,
#            compute_effective_envelope, MonotonicTighteningError
#
# Run: uv run python textbook/python/06-pact/04_envelopes.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

from datetime import UTC, datetime, timedelta

from kailash.trust import ConfidentialityLevel
from kailash.trust.pact import (
    MonotonicTighteningError,
    RoleEnvelope,
    TaskEnvelope,
    compute_effective_envelope,
    intersect_envelopes,
)
from kailash.trust.pact.config import (
    ConstraintEnvelopeConfig,
    DataAccessConstraintConfig,
    FinancialConstraintConfig,
    OperationalConstraintConfig,
)

# ── 1. Create a ConstraintEnvelopeConfig ─────────────────────────────
# The envelope has five CARE dimensions: financial, operational, temporal,
# data_access, communication. Plus confidentiality ceiling.

cto_envelope = ConstraintEnvelopeConfig(
    id="env-cto",
    description="CTO operating envelope",
    confidentiality_clearance=ConfidentialityLevel.SECRET,
    financial=FinancialConstraintConfig(
        max_spend_usd=50000.0,
        api_cost_budget_usd=5000.0,
        requires_approval_above_usd=10000.0,
    ),
    operational=OperationalConstraintConfig(
        allowed_actions=["read", "write", "deploy", "approve"],
        blocked_actions=["delete_production"],
    ),
    data_access=DataAccessConstraintConfig(
        read_paths=["/engineering/**", "/shared/**"],
        write_paths=["/engineering/**"],
    ),
)

assert cto_envelope.id == "env-cto"
assert cto_envelope.financial is not None
assert cto_envelope.financial.max_spend_usd == 50000.0
assert "deploy" in cto_envelope.operational.allowed_actions
assert "delete_production" in cto_envelope.operational.blocked_actions

# ── 2. Create a tighter envelope for a direct report ─────────────────

lead_envelope = ConstraintEnvelopeConfig(
    id="env-lead",
    description="Backend Lead operating envelope",
    confidentiality_clearance=ConfidentialityLevel.CONFIDENTIAL,
    financial=FinancialConstraintConfig(
        max_spend_usd=5000.0,
        api_cost_budget_usd=1000.0,
        requires_approval_above_usd=2000.0,
    ),
    operational=OperationalConstraintConfig(
        allowed_actions=["read", "write", "deploy"],
        blocked_actions=["delete_production", "approve"],
    ),
    data_access=DataAccessConstraintConfig(
        read_paths=["/engineering/**"],
        write_paths=["/engineering/backend/**"],
    ),
)

# Lead has lower spend, fewer actions, narrower paths — tighter in every dimension.

# ── 3. Intersect two envelopes ───────────────────────────────────────
# intersect_envelopes() computes the most restrictive combination.
# Financial: min() of numeric limits
# Operational: intersection of allowed, union of blocked
# Data access: intersection of paths

intersection = intersect_envelopes(cto_envelope, lead_envelope)

# Financial: min of spend limits
assert intersection.financial is not None
assert intersection.financial.max_spend_usd == 5000.0  # min(50000, 5000)
assert intersection.financial.api_cost_budget_usd == 1000.0  # min(5000, 1000)

# Operational: intersection of allowed, union of blocked
assert "read" in intersection.operational.allowed_actions
assert "write" in intersection.operational.allowed_actions
# "deploy" is in both allowed sets, so it stays
assert "deploy" in intersection.operational.allowed_actions
# "approve" is blocked by lead, so removed from allowed and added to blocked
assert "approve" not in intersection.operational.allowed_actions
assert "approve" in intersection.operational.blocked_actions
assert "delete_production" in intersection.operational.blocked_actions

# Data access: intersection of paths
assert "/engineering/**" in intersection.data_access.read_paths
# "/shared/**" only in CTO, not in lead, so excluded from intersection
assert "/shared/**" not in intersection.data_access.read_paths

# Confidentiality: min of levels
assert intersection.confidentiality_clearance == ConfidentialityLevel.CONFIDENTIAL

# ── 4. RoleEnvelope — standing boundary from supervisor ──────────────

role_env = RoleEnvelope(
    id="re-backend-lead",
    defining_role_address="D1-R1",  # CTO defines it
    target_role_address="D1-R1-T1-R1",  # for Backend Lead
    envelope=lead_envelope,
)

assert role_env.defining_role_address == "D1-R1"
assert role_env.target_role_address == "D1-R1-T1-R1"
assert role_env.version == 1

# RoleEnvelope is frozen
try:
    role_env.version = 2  # type: ignore[misc]
    assert False, "Should raise FrozenInstanceError"
except AttributeError:
    pass  # Expected: frozen=True

# ── 5. Monotonic tightening validation ───────────────────────────────
# Child envelopes must be at most as permissive as parent envelopes.

# This should pass — lead_envelope is tighter than cto_envelope
RoleEnvelope.validate_tightening(
    parent_envelope=cto_envelope,
    child_envelope=lead_envelope,
)

# This should fail — trying to give child more spend than parent
wider_envelope = ConstraintEnvelopeConfig(
    id="env-too-wide",
    financial=FinancialConstraintConfig(
        max_spend_usd=100000.0,  # More than CTO's 50000
    ),
    operational=OperationalConstraintConfig(
        allowed_actions=["read"],
    ),
)

try:
    RoleEnvelope.validate_tightening(
        parent_envelope=cto_envelope,
        child_envelope=wider_envelope,
    )
    assert False, "Should raise MonotonicTighteningError"
except MonotonicTighteningError:
    pass  # Expected: child max_spend exceeds parent

# ── 6. TaskEnvelope — ephemeral narrowing for a task ─────────────────

task_env = TaskEnvelope(
    id="te-deploy-v2",
    task_id="task-deploy-v2",
    parent_envelope_id="re-backend-lead",
    envelope=ConstraintEnvelopeConfig(
        id="env-deploy-task",
        confidentiality_clearance=ConfidentialityLevel.RESTRICTED,
        financial=FinancialConstraintConfig(
            max_spend_usd=500.0,  # Much tighter for this specific task
        ),
        operational=OperationalConstraintConfig(
            allowed_actions=["read", "deploy"],
        ),
        data_access=DataAccessConstraintConfig(
            read_paths=["/engineering/backend/**"],
            write_paths=["/engineering/backend/deployments/**"],
        ),
    ),
    expires_at=datetime.now(UTC) + timedelta(hours=4),
)

assert task_env.task_id == "task-deploy-v2"
assert task_env.is_expired is False

# Expired task envelope
expired_task = TaskEnvelope(
    id="te-old",
    task_id="task-old",
    parent_envelope_id="re-backend-lead",
    envelope=ConstraintEnvelopeConfig(
        id="env-old-task",
        operational=OperationalConstraintConfig(allowed_actions=["read"]),
    ),
    expires_at=datetime.now(UTC) - timedelta(hours=1),  # Already expired
)

assert expired_task.is_expired is True

# ── 7. compute_effective_envelope — three-layer model ────────────────
# Effective = intersection of all ancestor RoleEnvelopes + TaskEnvelope.
#
# RoleEnvelope (standing)
#   intersection (monotonic tightening)
# TaskEnvelope (ephemeral)
#   =
# EffectiveEnvelope (computed)

# Create supervisor envelope
supervisor_role_env = RoleEnvelope(
    id="re-cto",
    defining_role_address="R1",  # Board defines CTO envelope
    target_role_address="D1-R1",
    envelope=cto_envelope,
)

# Create direct report envelope
report_role_env = RoleEnvelope(
    id="re-lead",
    defining_role_address="D1-R1",  # CTO defines lead envelope
    target_role_address="D1-R1-T1-R1",
    envelope=lead_envelope,
)

# Map envelopes by target_role_address
role_envelopes = {
    "D1-R1": supervisor_role_env,
    "D1-R1-T1-R1": report_role_env,
}

# Compute effective envelope for the Backend Lead
effective = compute_effective_envelope(
    role_address="D1-R1-T1-R1",
    role_envelopes=role_envelopes,
    task_envelope=None,
)

assert effective is not None
# The effective envelope is the intersection of CTO + Lead envelopes
assert effective.financial is not None
assert effective.financial.max_spend_usd == 5000.0  # min(50000, 5000)

# ── 8. Effective envelope with task envelope ─────────────────────────

effective_with_task = compute_effective_envelope(
    role_address="D1-R1-T1-R1",
    role_envelopes=role_envelopes,
    task_envelope=task_env,
)

assert effective_with_task is not None
assert effective_with_task.financial is not None
assert effective_with_task.financial.max_spend_usd == 500.0  # min(5000, 500)

# ── 9. Expired task envelopes are ignored ────────────────────────────

effective_expired_task = compute_effective_envelope(
    role_address="D1-R1-T1-R1",
    role_envelopes=role_envelopes,
    task_envelope=expired_task,
)

assert effective_expired_task is not None
# Expired task is skipped, so result equals role-only effective
assert effective_expired_task.financial is not None
assert effective_expired_task.financial.max_spend_usd == 5000.0

# ── 10. No envelopes means maximally permissive ─────────────────────

no_envelope = compute_effective_envelope(
    role_address="D1-R1",
    role_envelopes={},
    task_envelope=None,
)
assert no_envelope is None  # None = no constraints (maximally permissive)

# ── 11. NaN/Inf validation — security invariant ─────────────────────
# NaN bypasses all numeric comparisons. Inf bypasses budget checks.
# The SDK rejects both explicitly.

try:
    FinancialConstraintConfig(
        max_spend_usd=float("nan"),
    )
    assert False, "Should reject NaN"
except ValueError:
    pass  # Expected: NaN values bypass governance checks

try:
    FinancialConstraintConfig(
        max_spend_usd=float("inf"),
    )
    assert False, "Should reject Inf"
except ValueError:
    pass  # Expected: Inf values bypass governance checks

# ── 12. Dimension scope for intersections ────────────────────────────
# When dimension_scope is provided, only those dimensions are intersected
# from the second envelope. Others are taken from the first unchanged.

base = ConstraintEnvelopeConfig(
    id="env-base",
    financial=FinancialConstraintConfig(max_spend_usd=10000.0),
    operational=OperationalConstraintConfig(
        allowed_actions=["read", "write", "deploy"],
    ),
)

override = ConstraintEnvelopeConfig(
    id="env-override",
    financial=FinancialConstraintConfig(max_spend_usd=1000.0),
    operational=OperationalConstraintConfig(
        allowed_actions=["read"],
    ),
)

# Only intersect financial, keep operational from base
scoped = intersect_envelopes(
    base,
    override,
    dimension_scope=frozenset({"financial"}),
)

assert scoped.financial is not None
assert scoped.financial.max_spend_usd == 1000.0  # Intersected
# Operational is preserved from base (not intersected)
assert "deploy" in scoped.operational.allowed_actions

print("PASS: 06-pact/04_envelopes")

File: textbook/rust/06-pact/src/bin/04_envelopes.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- PACT / Operating Envelopes
//!
//! OBJECTIVE: Define and enforce operating envelopes for AI agents.
//! LEVEL: Intermediate
//! PARITY: Full -- Both SDKs use Envelope with identical constraint semantics.
//! VALIDATES: Envelope, EnvelopeConstraint, budget limits, action restrictions
//!
//! Run: cargo run -p tutorial-pact --bin 04_envelopes

use kailash_governance::envelopes::{Envelope, EnvelopeConstraint};

fn main() {
    // ── 1. Operating Envelope ──
    // An envelope defines the boundaries within which an AI agent operates.
    // It constrains: budget, actions, data access, time, and escalation.

    let envelope = Envelope::new("code-reviewer")
        .budget_usd(5.0)
        .max_actions(100)
        .allowed_action("read_file")
        .allowed_action("search_code")
        .allowed_action("write_review")
        .denied_action("delete_file")
        .denied_action("push_to_main")
        .clearance_required(3);

    assert_eq!(envelope.name(), "code-reviewer");
    assert_eq!(envelope.budget_usd(), 5.0);
    assert_eq!(envelope.max_actions(), 100);

    // ── 2. Action enforcement ──
    // Check whether an action is permitted within the envelope.

    assert!(envelope.is_action_allowed("read_file"));
    assert!(envelope.is_action_allowed("search_code"));
    assert!(!envelope.is_action_allowed("delete_file"));
    assert!(!envelope.is_action_allowed("push_to_main"));

    // ── 3. Budget tracking ──
    // Envelopes track consumed budget.

    let mut tracked = envelope.clone();
    tracked.record_cost(2.0);
    assert_eq!(tracked.consumed_usd(), 2.0);
    assert_eq!(tracked.remaining_usd(), 3.0);
    assert!(!tracked.is_budget_exhausted());

    tracked.record_cost(3.5);
    assert!(tracked.is_budget_exhausted());

    // ── 4. Action counting ──
    // Track actions taken against the max_actions limit.

    let mut counted = envelope.clone();
    for _ in 0..50 {
        counted.record_action();
    }
    assert_eq!(counted.actions_taken(), 50);
    assert!(!counted.is_action_limit_reached());

    for _ in 0..50 {
        counted.record_action();
    }
    assert!(counted.is_action_limit_reached());

    // ── 5. EnvelopeConstraint ──
    // Individual constraints can be checked independently.

    let budget_constraint = EnvelopeConstraint::Budget { max_usd: 10.0 };
    let action_constraint = EnvelopeConstraint::MaxActions { limit: 50 };
    let clearance_constraint = EnvelopeConstraint::Clearance { min_level: 3 };

    assert!(matches!(budget_constraint, EnvelopeConstraint::Budget { .. }));
    assert!(matches!(action_constraint, EnvelopeConstraint::MaxActions { .. }));
    assert!(matches!(clearance_constraint, EnvelopeConstraint::Clearance { .. }));

    // ── 6. Nested envelopes ──
    // Sub-agents inherit parent envelope constraints.
    // The sub-agent's envelope must be within the parent's limits.

    let parent_envelope = Envelope::new("supervisor")
        .budget_usd(20.0)
        .max_actions(500);

    let child_envelope = Envelope::new("worker")
        .budget_usd(5.0) // Must be <= parent's 20.0
        .max_actions(100); // Must be <= parent's 500

    assert!(child_envelope.budget_usd() <= parent_envelope.budget_usd());
    assert!(child_envelope.max_actions() <= parent_envelope.max_actions());

    // ── 7. Key concepts ──
    // - Envelope: operating boundaries for AI agents
    // - Budget constraint: maximum USD spend
    // - Action constraint: maximum number of actions
    // - Action allow/deny lists: permitted operations
    // - Clearance requirement: minimum clearance level
    // - Budget and action tracking during execution
    // - Nested envelopes: child must be within parent limits
    // - Fail-closed: denied by default, explicitly allowed

    println!("PASS: 06-pact/04_envelopes");
}
